Tar up contents of child directory onto tmpfs if mounted over

rhatdan opened the pull request:
This patch will use the new PreMount and PostMount hooks to "tar"
up the contents of the base image on top of tmpfs mount points.

We want to add support for /run and /tmp on tmpfs and Michael Chrosby idea to
add --tmpfs PATH would require a patch like this.

Docker-DCO-1.1-Signed-off-by: Dan Walsh dwalsh@redhat.com (github: rhatdan)

rhatdan committed:
Tar up contents of child directory onto tmpfs if mounted over

This patch will use the new PreMount and PostMount hooks to "tar"
up the contents of the base image on top of tmpfs mount points.

We want to add support for /run and /tmp on tmpfs and Michael Chrosby idea to
add --tmpfs PATH would require a patch like this.

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

lheinlen-os commented:
I believe that this pull request is causing issues similar to the following when creating containers from images which do not have a /tmp directory (like an image created using `scratch` as the base).
```
[root@centos-7-docker ~]# docker run -ti hello
Timestamp: 2015-05-14 18:27:15.839329738 +0000 UTC
Code: System error

Message: [/usr/bin/tar -cf /var/lib/docker/tmp/de448245af0b33e5bb295979acdee7920a08382120c6db1d0b2de4b512d8a153664414103/_tmp.tar -C /var/lib/docker/devicemapper/mnt/de448245af0b33e5bb295979acdee7920a08382120c6db1d0b2de4b512d8a153/rootfs/tmp .] failed: /usr/bin/tar: /var/lib/docker/devicemapper/mnt/de448245af0b33e5bb295979acdee7920a08382120c6db1d0b2de4b512d8a153/rootfs/tmp: Cannot chdir: No such file or directory
/usr/bin/tar: Error is not recoverable: exiting now
: exit status 2
```

rhatdan commented:
@lheinlen-os Is that on a Fedora or RHEL system? 

This patch would not cause that but using the patch for mounting a tmpfs on /tmp might.

rhatdan commented:
@lheinlen-os I added a check to make sure fullDest exists before running the tar commands which should prevent the failure you are seeing

lheinlen-os commented:
@rhatdan Excellent!  Thank you!

rhatdan commented:
I am reworking these patches to only happen when  you run in systemd mode.  BTW.

rhatdan commented:
https://github.com/docker/docker/pull/13525
 uses this patch.

rhatdan committed:
Add tmpfs as a valid volume source command.

This patch will allow users to setup tmpfs mounts on top of
any directory.  One of the big benefits of this would be
to allow --read-only to work when an application needs to write to /run or /tmp.

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

rhatdan commented:
I have incorporated this pull request into https://github.com/docker/docker/pull/13587
